File: backend/app/routes/jobs.py
"""Job posting routes — public listings, admin CRUD, publish, archive."""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, Query, Request
from starlette.datastructures import UploadFile
from supabase import Client

from app.core.deps import get_current_user, get_supabase
from app.core.exceptions import NotFoundError, ValidationError
from app.core.security import require_role
from app.repositories.job_repo import JobRepo
from app.schemas.common import MessageResponse
from app.schemas.jobs import JobCreate, JobListResponse, JobResponse, JobUpdate, JdParseResponse
from app.services.job_service import JobService
from app.services.jd_service import parse_job_document, parse_job_description
from app.services.workspace_access import WorkspaceAccess, recruiter_actor
from app.voice.authorization import Actor, require_job

router = APIRouter(prefix="/jobs", tags=["jobs"])


def _job_service(supabase: Client = Depends(get_supabase)) -> JobService:
    return JobService(job_repo=JobRepo(supabase))


# ── Public endpoints (no auth) ──────────────────────────────


@router.get("/public", response_model=JobListResponse)
async def list_public_jobs(
    page: int = Query(1, ge=1),
    per_page: int = Query(20, ge=1, le=100),
    search: str | None = Query(None),
    department: str | None = Query(None),
    service: JobService = Depends(_job_service),
) -> JobListResponse:
    """List published job postings (public, no auth required)."""
    filters: dict[str, Any] = {}
    if search:
        filters["search"] = search
    if department:
        filters["department"] = department
    return await service.get_public_jobs(filters=filters, page=page, per_page=per_page)


@router.get("/public/{job_id}", response_model=JobResponse)
async def get_public_job(
    job_id: str,
    service: JobService = Depends(_job_service),
) -> JobResponse:
    """Get a single published job detail (public)."""
    job = await service.get_job(job_id)
    if job.status != "published":
        raise NotFoundError("Published job not found")
    return job


@router.post("/parse-jd", response_model=JdParseResponse)
async def parse_jd(
    request: Request,
    actor: Actor = Depends(recruiter_actor),
) -> JdParseResponse:
    """Parse raw JD text or an uploaded PDF/DOCX into a recruiter-editable preview."""
    content_type = (request.headers.get("content-type") or "").lower()
    if "multipart/form-data" in content_type:
        form = await request.form()
        upload = form.get("file")
        raw_text = form.get("text")
        if isinstance(upload, UploadFile):
            return parse_job_document(await upload.read(), upload.filename, upload.content_type)
        if isinstance(raw_text, str) and raw_text.strip():
            return parse_job_description(raw_text)
    elif "application/json" in content_type:
        payload = await request.json()
        raw_text = payload.get("text") if isinstance(payload, dict) else None
        if isinstance(raw_text, str) and raw_text.strip():
            return parse_job_description(raw_text)
    raise ValidationError("Provide JD text or a PDF/DOCX file")


# ── Admin / Recruiter endpoints ─────────────────────────────


@router.get("", response_model=JobListResponse)
async def list_jobs(
    page: int = Query(1, ge=1),
    per_page: int = Query(20, ge=1, le=100),
    status: str | None = Query(None),
    search: str | None = Query(None),
    department: str | None = Query(None),
    actor: Actor = Depends(recruiter_actor),
    service: JobService = Depends(_job_service),
    supabase: Client = Depends(get_supabase),
) -> JobListResponse:
    """List all jobs (admin/recruiter, all statuses)."""
    filters: dict[str, Any] = {}
    if status:
        filters["status"] = status
    if search:
        filters["search"] = search
    if department:
        filters["department"] = department
    filters["job_ids"] = await WorkspaceAccess(supabase, actor).job_ids()
    return await service.get_jobs(filters=filters, page=page, per_page=per_page)


@router.get("/{job_id}", response_model=JobResponse)
async def get_job(
    job_id: str,
    actor: Actor = Depends(recruiter_actor),
    service: JobService = Depends(_job_service),
    supabase: Client = Depends(get_supabase),
) -> JobResponse:
    """Get a single job for the recruiter workspace."""
    await require_job(actor, job_id, supabase)
    return await service.get_job(job_id)


@router.post("", response_model=JobResponse, status_code=201)
async def create_job(
    data: JobCreate,
    actor: Actor = Depends(recruiter_actor),
    service: JobService = Depends(_job_service),
) -> JobResponse:
    """Create a new job posting (draft)."""
    return await service.create_job(data, user_id=actor.user_id, tenant_id=actor.tenant_id)


@router.patch("/{job_id}", response_model=JobResponse)
async def update_job(
    job_id: str,
    data: JobUpdate,
    actor: Actor = Depends(recruiter_actor),
    service: JobService = Depends(_job_service),
    supabase: Client = Depends(get_supabase),
) -> JobResponse:
    """Update a job posting."""
    await require_job(actor, job_id, supabase)
    return await service.update_job(job_id, data)


@router.post("/{job_id}/publish", response_model=JobResponse)
async def publish_job(
    job_id: str,
    actor: Actor = Depends(recruiter_actor),
    service: JobService = Depends(_job_service),
    supabase: Client = Depends(get_supabase),
) -> JobResponse:
    """Publish a draft job posting."""
    await require_job(actor, job_id, supabase)
    return await service.publish_job(job_id)


@router.post("/{job_id}/archive", response_model=JobResponse)
async def archive_job(
    job_id: str,
    actor: Actor = Depends(recruiter_actor),
    service: JobService = Depends(_job_service),
    supabase: Client = Depends(get_supabase),
) -> JobResponse:
    """Archive a job posting."""
    await require_job(actor, job_id, supabase)
    return await service.archive_job(job_id)

File: frontend/src/lib/api/jobs.ts
/**
 * Intra AI — Jobs API Client Endpoints
 *
 * Verifiably mapped to backend/app/routes/jobs.py:
 * - GET /api/v1/jobs/public           (EXISTING, public)
 * - GET /api/v1/jobs/public/{job_id}  (EXISTING, public)
 * - GET /api/v1/jobs                  (EXISTING, admin/recruiter)
 * - POST /api/v1/jobs                 (EXISTING, admin/recruiter)
 * - PATCH /api/v1/jobs/{job_id}       (EXISTING, admin/recruiter)
 * - POST /api/v1/jobs/{job_id}/publish(EXISTING, admin/recruiter)
 * - POST /api/v1/jobs/{job_id}/archive(EXISTING, admin/recruiter)
 */

import { apiClient } from "./client";
import type {
  BackendJobResponse,
  BackendJobListResponse,
  BackendJobCreatePayload,
  BackendJobUpdatePayload,
  JdParseResponse,
  JdTextParsePayload,
} from "@/types/api";


export interface JobListParams {
  page?: number;
  per_page?: number;
  search?: string;
  department?: string;
  status?: string;
}

/**
 * Fetch public job listings (no auth required).
 * Backend: GET /api/v1/jobs/public
 */
export async function getPublicJobs(params?: JobListParams): Promise<BackendJobListResponse> {
  return apiClient<BackendJobListResponse>("/api/v1/jobs/public", {
    params: {
      page: params?.page,
      per_page: params?.per_page,
      search: params?.search,
      department: params?.department,
    },
    skipAuth: true,
  });
}

/**
 * Fetch a single public job by ID.
 * Backend: GET /api/v1/jobs/public/{job_id}
 */
export async function getPublicJob(jobId: string): Promise<BackendJobResponse> {
  return apiClient<BackendJobResponse>(`/api/v1/jobs/public/${jobId}`, {
    skipAuth: true,
  });
}

/**
 * Fetch a single job by ID (admin/recruiter).
 * Backend: GET /api/v1/jobs/{job_id}
 */
export async function getJob(jobId: string): Promise<BackendJobResponse> {
  return apiClient<BackendJobResponse>(`/api/v1/jobs/${jobId}`);
}

/**
 * Fetch all jobs for administrative dashboard (admin/recruiter only).
 * Backend: GET /api/v1/jobs
 */
export async function getJobs(params?: JobListParams): Promise<BackendJobListResponse> {
  return apiClient<BackendJobListResponse>("/api/v1/jobs", {
    params: {
      page: params?.page,
      per_page: params?.per_page,
      status: params?.status,
      search: params?.search,
      department: params?.department,
    },
  });
}

/**
 * Create a new draft job posting.
 * Backend: POST /api/v1/jobs
 */
export async function createJob(data: BackendJobCreatePayload): Promise<BackendJobResponse> {
  return apiClient<BackendJobResponse>("/api/v1/jobs", {
    method: "POST",
    body: JSON.stringify(data),
  });
}

/**
 * Update an existing job posting.
 * Backend: PATCH /api/v1/jobs/{job_id}
 */
export async function updateJob(
  jobId: string,
  data: BackendJobUpdatePayload
): Promise<BackendJobResponse> {
  return apiClient<BackendJobResponse>(`/api/v1/jobs/${jobId}`, {
    method: "PATCH",
    body: JSON.stringify(data),
  });
}

/**
 * Publish a draft job posting.
 * Backend: POST /api/v1/jobs/{job_id}/publish
 */
export async function publishJob(jobId: string): Promise<BackendJobResponse> {
  return apiClient<BackendJobResponse>(`/api/v1/jobs/${jobId}/publish`, {
    method: "POST",
  });
}

/**
 * Archive an active or draft job posting.
 * Backend: POST /api/v1/jobs/{job_id}/archive
 */
export async function archiveJob(jobId: string): Promise<BackendJobResponse> {
  return apiClient<BackendJobResponse>(`/api/v1/jobs/${jobId}/archive`, {
    method: "POST",
  });
}

/**
 * Parse an uploaded JD file (PDF or DOCX).
 * Backend: POST /api/v1/jobs/parse-jd
 */
export async function parseJdFile(file: File): Promise<JdParseResponse> {
  const formData = new FormData();
  formData.append("file", file);
  return apiClient<JdParseResponse>("/api/v1/jobs/parse-jd", {
    method: "POST",
    body: formData,
  });
}

/**
 * Parse raw JD text string.
 * Backend: POST /api/v1/jobs/parse-jd
 */
export async function parseJdText(payload: JdTextParsePayload): Promise<JdParseResponse> {
  return apiClient<JdParseResponse>("/api/v1/jobs/parse-jd", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

